File: app/src/main/java/com/kickstarter/viewmodels/TwoFactorViewModel.java
package com.kickstarter.viewmodels;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.rx.transformers.Transformers;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.apiresponses.AccessTokenEnvelope;
import com.kickstarter.services.apiresponses.ErrorEnvelope;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.TwoFactorActivity;

import androidx.annotation.NonNull;
import rx.Observable;
import rx.subjects.PublishSubject;

public interface TwoFactorViewModel {

  interface Inputs {
    /** Call when the 2FA code has been submitted. */
    void code(String __);

    /** Call when the log in button has been clicked. */
    void loginClick();

    /** Call when the resend button has been clicked. */
    void resendClick();
  }

  interface Outputs {
    /** Emits when submitting TFA code errored for an unknown reason. */
    Observable<Void> genericTfaError();

    /** Emits when TFA code was submitted. */
    Observable<Boolean> formSubmitting();

    /** Emits when TFA code submission has completed. */
    Observable<Boolean> formIsValid();

    /** Emits when resend code confirmation should be shown. */
    Observable<Void> showResendCodeConfirmation();

    /** Emits when a submitted TFA code does not match. */
    Observable<Void> tfaCodeMismatchError();

    /** Emits when submitting TFA code was successful. */
    Observable<Void> tfaSuccess();
  }

  final class ViewModel extends ActivityViewModel<TwoFactorActivity> implements Inputs, Outputs{
    private final ApiClientType client;
    private final CurrentUserType currentUser;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.currentUser = environment.currentUser();
      this.client = environment.apiClient();

      final Observable<String> email = intent()
        .map(i -> i.getStringExtra(IntentKey.EMAIL));
      final Observable<String> fbAccessToken = intent()
        .map(i -> i.getStringExtra(IntentKey.FACEBOOK_TOKEN));
      final Observable<Boolean> isFacebookLogin = intent()
        .map(i -> i.getBooleanExtra(IntentKey.FACEBOOK_LOGIN, false));
      final Observable<String> password = intent()
        .map(i -> i.getStringExtra(IntentKey.PASSWORD));

      final Observable<TfaData> tfaData = Observable.combineLatest(
        email, fbAccessToken, isFacebookLogin, password, TfaData::new
      );

      this.code
        .map(TwoFactorViewModel.ViewModel::isCodeValid)
        .compose(bindToLifecycle())
        .subscribe(this.formIsValid);

      this.code
        .compose(Transformers.combineLatestPair(tfaData))
        .compose(Transformers.takeWhen(this.loginClick))
        .filter(cd -> !cd.second.isFacebookLogin)
        .switchMap(cd -> this.login(cd.first, cd.second.email, cd.second.password))
        .compose(bindToLifecycle())
        .subscribe(this::success);

      this.code
        .compose(Transformers.combineLatestPair(tfaData))
        .compose(Transformers.takeWhen(this.loginClick))
        .filter(cd -> cd.second.isFacebookLogin)
        .switchMap(cd -> this.loginWithFacebook(cd.first, cd.second.fbAccessToken))
        .compose(bindToLifecycle())
        .subscribe(this::success);

      tfaData
        .compose(Transformers.takeWhen(this.resendClick))
        .filter(d -> !d.isFacebookLogin)
        .flatMap(d -> resendCode(d.email, d.password))
        .compose(bindToLifecycle())
        .subscribe();

      tfaData
        .compose(Transformers.takeWhen(this.resendClick))
        .filter(d -> d.isFacebookLogin)
        .flatMap(d -> resendCodeWithFacebook(d.fbAccessToken))
        .compose(bindToLifecycle())
        .subscribe();

      this.lake.trackTwoFactorConfirmationViewed();
    }

    private void success(final @NonNull AccessTokenEnvelope envelope) {
      this.currentUser.login(envelope.user(), envelope.accessToken());
      this.tfaSuccess.onNext(null);
    }

    private @NonNull Observable<AccessTokenEnvelope> login(final @NonNull String code, final @NonNull String email,
      final @NonNull String password) {
      return this.client.login(email, password, code)
        .compose(Transformers.pipeApiErrorsTo(this.tfaError))
        .compose(Transformers.neverError())
        .doOnSubscribe(() -> this.formSubmitting.onNext(true))
        .doAfterTerminate(() -> this.formSubmitting.onNext(false));
    }

    private @NonNull Observable<AccessTokenEnvelope> loginWithFacebook(final @NonNull String code, final @NonNull String fbAccessToken) {
      return this.client.loginWithFacebook(fbAccessToken, code)
        .compose(Transformers.pipeApiErrorsTo(this.tfaError))
        .compose(Transformers.neverError())
        .doOnSubscribe(() -> this.formSubmitting.onNext(true))
        .doAfterTerminate(() -> this.formSubmitting.onNext(false));
    }

    private @NonNull Observable<AccessTokenEnvelope> resendCode(final @NonNull String email, final @NonNull String password) {
      return this.client.login(email, password)
        .compose(Transformers.neverError())
        .doOnSubscribe(() -> this.showResendCodeConfirmation.onNext(null));
    }

    private @NonNull Observable<AccessTokenEnvelope> resendCodeWithFacebook(final @NonNull String fbAccessToken) {
      return this.client.loginWithFacebook(fbAccessToken)
        .compose(Transformers.neverError())
        .doOnSubscribe(() -> this.showResendCodeConfirmation.onNext(null));
    }

    private static boolean isCodeValid(final String code) {
      return code != null && code.length() > 0;
    }

    private final PublishSubject<String> code = PublishSubject.create();
    private final PublishSubject<Void> loginClick = PublishSubject.create();
    private final PublishSubject<Void> resendClick = PublishSubject.create();

    private final PublishSubject<Boolean> formIsValid = PublishSubject.create();
    private final PublishSubject<Boolean> formSubmitting = PublishSubject.create();
    private final PublishSubject<Void> showResendCodeConfirmation = PublishSubject.create();
    private final PublishSubject<ErrorEnvelope> tfaError = PublishSubject.create();
    private final PublishSubject<Void> tfaSuccess = PublishSubject.create();

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public Observable<Boolean> formIsValid() {
      return this.formIsValid;
    }
    @Override public Observable<Boolean> formSubmitting() {
      return this.formSubmitting;
    }
    @Override public Observable<Void> genericTfaError() {
      return this.tfaError
        .filter(env -> !env.isTfaFailedError())
        .map(__ -> null);
    }
    @Override public Observable<Void> showResendCodeConfirmation() {
      return this.showResendCodeConfirmation;
    }
    @Override public Observable<Void> tfaCodeMismatchError() {
      return this.tfaError
        .filter(ErrorEnvelope::isTfaFailedError)
        .map(__ -> null);
    }
    @Override public Observable<Void> tfaSuccess() {
      return this.tfaSuccess;
    }

    @Override public void code(final @NonNull String s) {
      this.code.onNext(s);
    }
    @Override public void loginClick() {
      this.loginClick.onNext(null);
    }
    @Override public void resendClick() {
      this.resendClick.onNext(null);
    }

    protected final class TfaData {
      final @NonNull String email;
      final @NonNull String fbAccessToken;
      final boolean isFacebookLogin;
      final @NonNull String password;

      protected TfaData(final @NonNull String email, final @NonNull String fbAccessToken, final boolean isFacebookLogin,
        final @NonNull String password) {
        this.email = email;
        this.fbAccessToken = fbAccessToken;
        this.isFacebookLogin = isFacebookLogin;
        this.password = password;
      }
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/TwoFactorViewModel.kt
package com.kickstarter.viewmodels

import android.content.Intent
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.ObjectUtils
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.models.User
import com.kickstarter.services.ApiClientTypeV2
import com.kickstarter.services.apiresponses.AccessTokenEnvelope
import com.kickstarter.services.apiresponses.ErrorEnvelope
import com.kickstarter.ui.IntentKey
import com.kickstarter.viewmodels.usecases.LoginUseCase
import com.kickstarter.viewmodels.usecases.RefreshUserUseCase
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

interface TwoFactorViewModel {
    interface Inputs {
        /** Call when the 2FA code has been submitted.  */
        fun code(code: String)

        /** Call when the log in button has been clicked.  */
        fun loginClick()

        /** Call when the resend button has been clicked.  */
        fun resendClick()
    }

    interface Outputs {
        /** Emits when submitting TFA code errored for an unknown reason.  */
        fun genericTfaError(): Observable<Unit>

        /** Emits when TFA code was submitted.  */
        fun formSubmitting(): Observable<Boolean>

        /** Emits when TFA code submission has completed.  */
        fun formIsValid(): Observable<Boolean>

        /** Emits when resend code confirmation should be shown.  */
        fun showResendCodeConfirmation(): Observable<Unit>

        /** Emits when a submitted TFA code does not match.  */
        fun tfaCodeMismatchError(): Observable<Unit>

        /** Emits when submitting TFA code was successful.  */
        fun tfaSuccess(): Observable<Unit>
    }

    class TwoFactorViewModel(
        environment: Environment,
        intent: Intent? = null
    ) : ViewModel(), Inputs, Outputs {
        private val client: ApiClientTypeV2 = requireNotNull(environment.apiClientV2())
        private val analytics = requireNotNull(environment.analytics())
        private val loginUserCase = LoginUseCase(environment)
        private val refreshUserUseCase = RefreshUserUseCase(environment)
        private val internalIntent = BehaviorSubject.createDefault(intent)
        private val disposables = CompositeDisposable()

        private fun success(user: User) {
            refreshUserUseCase.refresh(user)
            tfaSuccess.onNext(Unit)
        }

        private fun login(
            code: String,
            email: String,
            password: String
        ): Observable<AccessTokenEnvelope> {
            return client.login(email, password, code)
                .compose(Transformers.pipeApiErrorsToV2(tfaError))
                .compose(Transformers.neverErrorV2())
                .doOnSubscribe { formSubmitting.onNext(true) }
                .doAfterTerminate { formSubmitting.onNext(false) }
        }

        private fun loginWithFacebook(
            code: String,
            fbAccessToken: String
        ): Observable<AccessTokenEnvelope> {
            return client.loginWithFacebook(fbAccessToken, code)
                .compose(Transformers.pipeApiErrorsToV2(tfaError))
                .compose(Transformers.neverErrorV2())
                .doOnSubscribe { formSubmitting.onNext(true) }
                .doAfterTerminate { formSubmitting.onNext(false) }
        }

        private fun resendCode(email: String, password: String): Observable<AccessTokenEnvelope> {
            return client.login(email, password)
                .compose(Transformers.neverErrorV2())
                .doOnSubscribe { showResendCodeConfirmation.onNext(Unit) }
        }

        private fun resendCodeWithFacebook(fbAccessToken: String): Observable<AccessTokenEnvelope> {
            return client.loginWithFacebook(fbAccessToken)
                .compose(Transformers.neverErrorV2())
                .doOnSubscribe { showResendCodeConfirmation.onNext(Unit) }
        }

        private val code = PublishSubject.create<String>()
        private val loginClick = PublishSubject.create<Unit>()
        private val resendClick = PublishSubject.create<Unit>()
        private val formIsValid = PublishSubject.create<Boolean>()
        private val formSubmitting = PublishSubject.create<Boolean>()
        private val showResendCodeConfirmation = PublishSubject.create<Unit>()
        private val tfaError = PublishSubject.create<ErrorEnvelope>()
        private val tfaSuccess = PublishSubject.create<Unit>()

        val inputs: Inputs = this
        val outputs: Outputs = this

        override fun formIsValid(): Observable<Boolean> = formIsValid

        override fun formSubmitting(): Observable<Boolean> = formSubmitting

        override fun genericTfaError(): Observable<Unit> {
            return tfaError
                .filter { env: ErrorEnvelope -> !env.isTfaFailedError }
                .map { Unit }
        }

        override fun showResendCodeConfirmation(): Observable<Unit> = showResendCodeConfirmation

        override fun tfaCodeMismatchError(): Observable<Unit> {
            return tfaError
                .filter(ErrorEnvelope::isTfaFailedError)
                .map { Unit }
        }

        override fun tfaSuccess(): Observable<Unit> {
            return tfaSuccess
        }

        override fun code(s: String) {
            this.code.onNext(s)
        }

        override fun loginClick() {
            loginClick.onNext(Unit)
        }

        override fun resendClick() {
            resendClick.onNext(Unit)
        }

        protected inner class TfaData(
            val email: String,
            val isFacebookLogin: Boolean,
            val password: String
        )

        protected inner class TfaDataForFacebook(
            val fbAccessToken: String,
            val isFacebookLogin: Boolean
        )
        companion object {
            private fun isCodeValid(code: String?): Boolean {
                return code != null && code.isNotEmpty()
            }
        }

        init {

            val email = internalIntent
                .map { it.getStringExtra(IntentKey.EMAIL) }
                .filter { ObjectUtils.isNotNull(it) }

            val fbAccessToken = internalIntent
                .map { it.getStringExtra(IntentKey.FACEBOOK_TOKEN) }

            val isFacebookLogin = internalIntent
                .map { it.getBooleanExtra(IntentKey.FACEBOOK_LOGIN, false) }

            val password = internalIntent
                .map { it.getStringExtra(IntentKey.PASSWORD) }
                .filter { ObjectUtils.isNotNull(it) }

            val tfaData = Observable.combineLatest(
                email,
                isFacebookLogin,
                password
            ) { email: String, isFacebookLogin: Boolean, password: String ->
                TfaData(
                    email,
                    isFacebookLogin,
                    password
                )
            }

            val tfaFacebookData = Observable.combineLatest(
                fbAccessToken,
                isFacebookLogin
            ) { fbAccessToken: String?, isFacebookLogin: Boolean ->
                TfaDataForFacebook(
                    fbAccessToken = fbAccessToken ?: "",
                    isFacebookLogin = isFacebookLogin
                )
            }

            this.code
                .map { code: String? -> isCodeValid(code) }
                .subscribe { formIsValid.onNext(it) }
                .addToDisposable(disposables)


            this.code
                .compose(Transformers.combineLatestPair(tfaData))
                .compose(Transformers.takeWhenV2(loginClick))
                .filter { !it.second.isFacebookLogin }
                .switchMap {
                    login(
                        it.first,
                        it.second.email,
                        it.second.password
                    )
                }
                .switchMap {
                    this.loginUserCase
                        .loginAndUpdateUserPrivacyV2(it.user(), it.accessToken())
                }
                .subscribe { success(it) }
                .addToDisposable(disposables)

            this.code
                .compose(Transformers.combineLatestPair(tfaFacebookData))
                .compose(Transformers.takeWhenV2(loginClick))
                .filter { it.second.isFacebookLogin }
                .switchMap {
                    loginWithFacebook(
                        it.first,
                        it.second.fbAccessToken
                    )
                }
                .switchMap {
                    this.loginUserCase
                        .loginAndUpdateUserPrivacyV2(it.user(), it.accessToken())
                }
                .subscribe { success(it) }
                .addToDisposable(disposables)

            tfaData
                .compose(Transformers.takeWhenV2(resendClick))
                .filter { !it.isFacebookLogin }
                .flatMap {
                    resendCode(it.email, it.password)
                }
                .subscribe()
                .addToDisposable(disposables)

            tfaFacebookData
                .compose(Transformers.takeWhenV2(resendClick))
                .filter { it.isFacebookLogin }
                .flatMap {
                    resendCodeWithFacebook(it.fbAccessToken)
                }
                .subscribe()
                .addToDisposable(disposables)

            analytics.trackTwoFactorAuthPageViewed()
        }
    }

    class Factory(private val environment: Environment) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return TwoFactorViewModel(environment) as T
        }
    }
}
